### ISSUE TITLE
Topic Requests

### ISSUE THREAD

**Issue by piyanatk** [prickly-pythons/prickly-pythons#9]
Is there any topic that you want to learn? Please response here.

**Comment by johndwest** [comment]
multi-threaded programming

**Comment by kpolsen** [comment]
DataFrames from the Pandas package!

**Comment by maitrayeebose** [comment]
fits to different distributions, learn tricks from the experts
:)

**Comment by rjansen15** [comment]
What little utilities can you run (monolithic scripts or executables) in Python that do something  I cannot easily do in C-Shell, Fortran, IDL, SuperMongo ?

**Comment by ehsan112233** [comment]
1. getting an spectrum and calculating the area under each peak
2. reading 2D and 3D array file and make math on them

**Comment by piyanatk** [comment]
@maitrayeebose what kind of data do you want to fit? what functional form?

**Comment by ehsan112233** [comment]
What is the difference between Spyder and Anaconda?
Which one is faster? which one is the best for beginners?...

**Comment by maitrayeebose** [comment]
@piyanatk I work with a mass spectrometer, and I basically want to be able to make a program that analyzes and plots the data in all useful forms. The analysis requires me to fit linear and non-linear profiles and basically I would like to use python for it. It might be too simplistic but I want to see what else I can do.

**Comment by piyanatk** [comment]
Creating virtual machine in Anaconda

**Comment by interstellar1992** [comment]
@piyanatk Are there any good tutorials on python that you could recommend to me? I used python for about one year but I consider myself as a beginner as I only know/have used very basic staff in python . In addition to the tutorials, I am especially interested in plotting figures using python... I work on observational astronomy. Thanks!

**Comment by astroseth** [comment]
I'd like to know simple i/o. Read in a file of numbers, process them and plot them.

**Comment by kpolsen** [comment]
@astroseth,@exopython: Your requests sounds like the reasons why I want to learn more about DataFrames. It seems like a really smart way to deal with big arrays.

**Comment by kpolsen** [comment]
I'm thinking of sending out a poll for our next face-to-face meeting next week! Both for the topic and the exact time but with few options:

![screen shot 2015-10-01 at 21 06 19](https://cloud.githubusercontent.com/assets/13789517/10239292/53cd3c28-6880-11e5-9313-9646447f7799.png)

What other 'start-up' topics\* are people interested in?
- Maybe the multi-threaded programming lies a few weeks ahead, like next time?

**Comment by piyanatk** [comment]
@kpolsen I agree that parallel programing should be later, and I can do that topic.

**Comment by kpolsen** [comment]
The following topics were proposed at today's meeting (intro to packages):
Numpy (scipy)
Astropy
Matplotlib: object oriented vs functional programming?

**Comment by kpolsen** [comment]
Another topic: spectroscopic analysis using [PySpecKit](http://pyspeckit.bitbucket.org/html/sphinx/index.html)
If anybody has experience, he/she would be very welcome to present it.

**Comment by kpolsen** [comment]
[yt-project](http://yt-project.org/doc/index.html): A tool particularly well suited for visualization of simulation data. I just joined the group on Bitbucket and Slack and they've got some cool stuff!! Someone with experience could give a short presentation - actually one of the more active members, Cameron Hummels, (who is now at Caltech but previously was in Tucson) let me know that he could stop by later this semester.

**Comment by kpolsen** [comment]
Hey there - anybody interested in giving a short presentation next Thurs on matplotlib - simple plotting? Preferably object-oriented programming. @godber @piyanatk @johndwest @bajoshi ?

**Comment by godber** [comment]
@kpolsen I can't commit to that right now, we have Mastcam-Z PDR on the 27th and it will be demanding much of my time between now and then.

**Comment by johndwest** [comment]
@kpolsen, my matplotlib work is not particularly object-oriented. I'm looking forward to someone else giving a matplotlib OO presentation.

**Comment by godber** [comment]
I can do this sometime in November.

**Comment by kpolsen** [comment]
@johndwest: that's fine, maybe you can talk about some matplotlib basics?  And I could introduce astropy - I just used it for fits files, but I think that could be useful for some. What do you think guys?
@Godber: Yeah that would be great if you could do object-oriented plotting later!

**Comment by johndwest** [comment]
@kpolsen I could do that. When/where is the meeting this time? I have a 1pm with my advisor in ISTB-4.

**Comment by kpolsen** [comment]
@johndwest : I just booked PSF 566 on Thurs at 12-2pm, so if you be there in the beginning that'd be great! I will post on our homepage shortly, http://sese-python-users.github.io/

**Comment by johndwest** [comment]
OK by me.

**Comment by kpolsen** [comment]
Perfect! Next meeting will be 12-2pm in PSF 566 on Thurs, see also notice here: http://sese-python-users.github.io/ 
There will be no email notification this time, so please pass it on to anyone interested!

**Comment by kpolsen** [comment]
Another topic that we might want to discuss: Subprocess management (https://docs.python.org/3.5/library/subprocess.html) which is the art of starting a program OUTSIDE your python script and maybe using it's output for the rest of your script. I recently started using it to run cloudy models (http://www.nublado.org/) from python.

**Comment by kpolsen** [comment]
For next time! Can I get anybody to do a short presentation on one of the following on Thurs next week?:
- matplotlib object-oriented plotting ( @godber  @piyanatk ?)
- integration of a function ( @amloren1 ?)
- subprocess command (probably me, if nobody else?)

**Comment by godber** [comment]
I can do matplotlib.  1.5.0 came out just recently too.

**Comment by kpolsen** [comment]
Yes! Thank you @godber. About the version of matplotlib, if some of you want to upgrade, this is one way that works on Mac if you have pip installed:

```
pip  install --upgrade matplotlib
```

If someone can also show an example of integrating a function that would be great!

**Comment by kpolsen** [comment]
Oh, so would 11am work for you, @godber ? I'm trying to switch between starting at 11am and 12pm and going for 2hrs to include as many as possible.

**Comment by godber** [comment]
11AM would work.

**Comment by kpolsen** [comment]
Ok perfect, thank you so much, Austin. So next time is now officially Thursday at 11am in PSF 566, see also notice here: http://sese-python-users.github.io/

**Comment by Jansport1701** [comment]
I would like to learn more about anaconda and mathplot.

mike

**Comment by amloren1** [comment]
Anyone have experience using one of the multi-core processing modules with
something like pexpect? Basically looking to have multiple terminals open
running command line applications. I use linux but anyone can chime in with
a module recommendation. Thanks ya'll!

-Alejandro

**Comment by kpolsen** [comment]
Now that we talked a lot about plotting, what other topic that people want to hear about? Maybe we are ready for something more challenging! Here are ideas:
- multi-core/threading programming ( @amloren1 @piyanatk @johndwest )
- integration of a function ( @abhijithrajan @KPerry0505 )
- functional fits to distributions!
  I mentioned some people - if any of you are interested in presenting an applied example, please let me know. Next time is on Thursday Nov 19, probably noon

**Comment by ehsan112233** [comment]
Hi Karen,

I prefer the integration of a function.

Thanks,
Ehsan

Sent from my iPad

On Nov 13, 2015, at 3:19 PM, Karen Pardos Olsen notifications@github.com
wrote:

Now that we talked a lot about plotting, what other topic that people want
to hear about? Maybe we are ready for something more challenging! Here are
ideas:
- multi-core/threading programming ( @amloren1
  https://github.com/amloren1 @piyanatk https://github.com/piyanatk
  @johndwest https://github.com/johndwest )
- integration of a function ( @abhijithrajan
  https://github.com/abhijithrajan @KPerry0505
  https://github.com/KPerry0505 ) I mentioned some people - if any of
  you are interested in presenting an applied example, please let me know.
  Next time is on Thursday Nov 19, probably noon

**Comment by kpolsen** [comment]
Tomorrows meeting will be at 11am and only 1 hour because it seems people are very busy this week (and the room is booked at noon). I will try to prepare something on integrating a function, and if anybody else wants to share some specific code they're more than welcome! Otherwise, we'll just have an open meeting where people can sit and code python :)

**Comment by kpolsen** [comment]
On this Thursday (Dec 3), I'm in New Mexico, so if someone else wants to lead the meeting OR some of you just want to meet up and code together, let me know or book the room (PSF 566) directly from Evelyn.Barnes@asu.edu !

**Comment by kpolsen** [comment]
@sese-python-users/users : Guys and girls - last meeting this year! On this coming Thursday Dec 10th: what should we talk about? Any topics in particular that you feel we haven't touched upon yet? Should we end up hardcore with some multi-processing discussion?

**Comment by kpolsen** [comment]
@sese-python-users/users : It's 2016 and time to learn new stuff in Python. Let's start up next week (Jan 25-29) - I will send a doodle out with times for those interested, but probably Monday or Tuesday.
What do you want to learn? I spoke with a few people, and got the following ideas:
- (more) fitting to different functions
- plotting jpeg/png image in correct coordinates within python plotting window
- read in data (genfromtext, read_table and other ways)
- make movies
- run Monte Carlo simulations
- anaconda (this is my idea: I would like to talk about anaconda for installing and running python, if anybody is interested, particularly those of you getting started with python.)

... got more ideas? Post them here! And let me know if you are super good at one of the above and would like to give examples.

**Comment by amloren1** [comment]
Graphics packages would be a fun topic to go over. As well as gui making
with Tkinter. It's native and relatively easy to grasp.

**Comment by earnric** [comment]
Matplotlib and especially combining multiple subplots of different types in
a single figure.

**Comment by kpolsen** [comment]
@amloren1 : Do you think you could give a short intro to Tkinter next week? And if people are interested, we'll have a full session later on that  - it sounds fun!
@earnric : complicated subplots huh? That's useful and could maybe be combined with the request above to plot jpeg/png images INSIDE the python axes. Someone who would like to demonstrate this?

**Comment by kpolsen** [comment]
More requests that came in: Dealing with SQL databases in python, and visualization packages such as vispy (http://vispy.org/) and the yt project (http://yt-project.org/) of which I have some experience with yt.

**Comment by KPerry0505** [comment]
Yes! I love the idea of learning about SQL databases. I'm trying to expand my programming knowledge and would love to learn more about SQL. 

```
On Wednesday, January 20, 2016 2:34 PM, Karen Pardos Olsen <notifications@github.com> wrote:
```

 Another request that came in: Dealing with SQL databases in python, and visualization packages such as vispy (http://vispy.org/) and the yt project (http://yt-project.org/) of which I have some experience with yt.—

**Comment by amloren1** [comment]
Karen, I can give a small intro to the tkinter module. When are the
meetings this semester?

**Comment by SeanBryan** [comment]
For python and SQL, I'm working on this tutorial now, once I find time to
actually finish it, I could give a talk about my experience.

https://plot.ly/ipython-notebooks/big-data-analytics-with-pandas-and-sqlite/

-Sean

**Comment by kpolsen** [comment]
@amloren1: Great! The time of the meetings has not been settled yet, but I'm about to send an email out with a doodle for a time and a spreadsheet to get an overview of people in the group.
@SeanBryan : sounds good, let's find a strange database to use for testing :)

**Comment by johndwest** [comment]
I do a lot of stuff with python and PostgreSQL, happy to talk about it.

```
 -- John
```

**Comment by spacegal-spiff** [comment]
Ditto re: python interfacing with SQL and databases! I would love to hear how that tutorial goes, @SeanBryan! Incidentally, Codecademy recently released a basic SQL learning module a few months ago that looks good: https://www.codecademy.com/learn/learn-sql

Also, I've been working with AstroPy's aperture photometry and source extraction utilities recently (specifically <a href="https://photutils.readthedocs.org/en/latest/">photutils</a>), so if anyone else is interested in photometry and PSF fitting, I wouldn't mind discussing it later in the semester. :)

**Comment by kpolsen** [comment]
Ok guys, how about this:
- I talk a bit about anaconda and subplotting (revisited)
- @amloren1 introduce the Tkinter module - do you have time?
  ... looks like it's going to be Monday 4pm!
  And then we'll have a SQL-focused session next time!

**Comment by kpolsen** [comment]
You should be receiving an email that the above is happening on Monday at 4pm in GW 505. As before, future meetings will also be announced here: http://prickly-pythons.github.io/ - hope you can all make it, there will be girl scouts cookies!

**Comment by amloren1** [comment]
hey Karen, I can talk about tkinter and get everyone going on making simple
guis

-amloren1

**Comment by earnric** [comment]
If you guys are gonna talk about subplotting maybe you can help me with one I need for my paper. I have these 2D histograms that I render with `pcolormesh`, but I haven't been able to get them to share a common x-axis... Here's a sample of the code. The topmost plot also has an x-axis value histogram.

```
X, Y = np.meshgrid(xrange, yrange)  # Create a grid over the range of bins for the plot = "[O/H]"
H, xedges, yedges = np.histogram2d((Mg_mixedF60/Mgsun)/(Ca_mixedF60/Casun), (Fe_mixedF60/Fesun), weights=masses[mixedStar] * (1.0 - spPF[mixedStar]), bins=(yrange,xrange))

fig = plt.figure(figsize=(18, 15))
axHistx = plt.axes(rect_histx)
ax2dhist = plt.axes(rect_2dhist)

cfeZero,  = ax2dhist.plot(np.array([1e-8,1e-2]),np.array([1,1]), 'k--')

# Fix colors -- white for values of 1.0. 
cmap = copy.copy(mpl.cm.jet)
cmap.set_bad('w', 1.)  # w is color, for values of 1.0
# Create a plot of the binned data, use a fixed lognormal scale
cax = (ax2dhist.pcolormesh(X, Y, H, cmap=cmap, norm=LogNorm(vmin=cbarMin,vmax=cbarMax))) # pcolormesh transposes X,Y
cbarticks = [1,2,4,6,cbarMax]
cbar = fig.colorbar(cax, ticks=cbarticks)
cbar.ax.set_yticklabels(cbarticks, size=26)
cbar.set_label("log $(M\, / d\,[Fe/H]\, / d\,"+yLabel+"\, /\, Mpc^{3})$ ", size=34)

N, bins, patches = axHistx.hist(Fe_mixedF60/Fesun, bins=xrange, log=True, 
                            weights=masses[mixedStar] * (1.0 - spPF[mixedStar]))
```

Here's what the plots currently look like:

![image](https://cloud.githubusercontent.com/assets/8357712/12534687/4df61b3a-c221-11e5-8d51-3821ca6bdd7a.png)

**Comment by kpolsen** [comment]
@amloren1 : thanks a lot, looking forward to hearing about Tkinter.
@earnric : that's a very interesting example that I think we should talk about today. What exactly are you missing in your plots? To remove the whitespace in between histogram and colormesh at the top? My initial guess would be to use subplots in some way: http://matplotlib.org/examples/pylab_examples/shared_axis_demo.html but maybe you tried that?

**Comment by earnric** [comment]
@kpolsen Yea, I've looked at that demo, but I'm having trouble setting up the 4 plots 1 of which is a different size. Right now I'm rendering 3 separate plots and just pasting them together, manually. (The top plot does include the histogram across the top of the x-axis... I got that working at least!)

**Comment by bajoshi** [comment]
@earnric -- have you looked at gridspec? I usually use gridspec anytime I have subplots that are different sizes and when I also want to mess with the white space between them.

Here's a short snippet of code that I have used before which has subplots with varying sizes and you can change the white space with the wspace and hspace keywords.

import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec

gs1 = gridspec.GridSpec(4,4)
gs1.update(left=0.05, right=0.95, bottom=0.30, top=0.99, wspace=0.05)
ax1 = plt.subplot(gs1[:-2, :2])
ax2 = plt.subplot(gs1[:-2, 2:])
ax3 = plt.subplot(gs1[2:, :2])
ax4 = plt.subplot(gs1[2:, 2:])

gs2 = gridspec.GridSpec(1,3)
gs2.update(left=0.05, right=0.95, bottom=0.01, top=0.25, wspace=0.05)
ax5 = plt.subplot(gs2[:, :1])
ax6 = plt.subplot(gs2[:, 1:2])
ax7 = plt.subplot(gs2[:, 2:3])

plt.show()

**Comment by kpolsen** [comment]
Thanks for a great meeting and presentation (@amloren1) today! 
In two weeks I can't make it, but you guys (@SeanBryan, @johndwest , @KPerry0505 ) should get together and do a SQL-session! I'll leave the left-over cookies for you.
Remember to put your name and interests here as well:
https://docs.google.com/spreadsheets/d/199XG0i2m-oL-e5lQPdnuVfxHZNo70Jtxxm_BhLhln0I/edit#gid=186001220

**Comment by johndwest** [comment]
I won't be able to make meetings for a while, I have physical therapy
sessions that overlap with the time slot. Bummer.

```
 -- John
```

**Comment by kpolsen** [comment]
@SeanBryan and @johndwest : Are you guys (or one of you, I know Sean is interested) available to talk about SQL on Monday at 4pm? I won't be there, so you would have to start up the meeting in room 505, but not much work is needed, just a 10-15min walkthrough of an example and people can continue on their own.

**Comment by johndwest** [comment]
I won't be at meetings for a while; I have physical therapy overlapping
that time slot. Sorry.

```
 -- John
```

**Comment by kpolsen** [comment]
Oh right. The other time that got a lot of votes was noon, would that be better? If not, I say we stick to the 4pm time. I will send an email out later today to seseall about what we decide.

**Comment by johndwest** [comment]
I'd say stick with 4pm. My PT schedule is subject to change.

  -- John

**Comment by kpolsen** [comment]
@prickly-pythons/users : Thanks guys (Sean in particular) for having the session on SQL this past Monday. Remember to post any topics you would like to discuss here, next time is Feb 22.
@amloren1 and @SeanBryan, if you find the time, please add example python scripts from your talks!

**Comment by godber** [comment]
FYI there is a new abstraction layer on top of SQL databases for python which might simplify things for users new to relational databases: https://github.com/kennethreitz/records.  I didn't see @SeanBryan's presentation, maybe he included it, but maybe not, **records** is pretty new.

**Comment by kpolsen** [comment]
@prickly-pythons/users : OBS! No meeting this Monday (22) as I am still out of town and no requests have come in. But let's meet the following **Monday (29) at 4 pm** and talk about:
- different ways of reading in data. If you have favorite methods, please share them with the rest of us on Monday.
- visualization of data with the yt project, which I have some experience with, but help is most welcomed!
- other? 
  think about what you would like to be better at in python and take advantage of the meetings!

**Comment by kpolsen** [comment]
@prickly-pythons/users : Who would be interested in talking about Monte Carlo simulations on next Monday (the 14th)? I'm thinking of maybe @godber , @piyanatk, @abhijithrajan or Enrico (who was the first to suggets MC)? I can 90% promise there'll be baked goods!

**Comment by kpolsen** [comment]
Topic ideas from todays meeting:
- calling scripts from other scripts, organizing your code
- importing data (ascii, txt, fits...)

... post your ideas here!

**Comment by kpolsen** [comment]
Another topic has been proposed: Making your website! Not really python-related but important in many other ways! And since we don't have a curriculum, we can talk about computer-related stuff outside of python if you guys are interested in specific topics.
